src/intrinsics.cpp

```diff
@@ -162,6 +279,127 @@ Symbol *lCreateISPCSymbolForLLVMIntrinsic(llvm::Function *func, SymbolTable *sym
     return sym;
 }
 
+static ISPCIntrinsics lLookupISPCInstrinsic(const std::string &name) {
+    llvm::StringRef ref(name);
+    if (ref.starts_with("llvm.ispc.atomicrmw.")) {
+        return ISPCIntrinsics::atomicrmw;
+    }
+    if (ref.starts_with("llvm.ispc.cmpxchg.")) {
+        return ISPCIntrinsics::atomicrmw;
+    }
+    if (ref.starts_with("llvm.ispc.fence.")) {
+        return ISPCIntrinsics::fence;
+    }
+    for (unsigned i = 1; i <= (unsigned)ISPCIntrinsics::stream_store; i++) {
+        if (name == ISPCIntrinsicsNames[i]) {
+            return (ISPCIntrinsics)i;
+        }
+    }
+    return ISPCIntrinsics::not_intrinsic;
+}
+
+static llvm::VectorType *lGetDoubleWidthVectorType(llvm::Type *type) {
+    llvm::VectorType *vt = llvm::dyn_cast<llvm::VectorType>(type);
+    if (vt == nullptr) {
+        Error(SourcePos(), "Expected vector type.");
+        return nullptr;
+    }
+    unsigned int vectorWidth = vt->getElementCount().getKnownMinValue();
+    return llvm::VectorType::get(vt->getElementType(), vectorWidth * 2, false);
+}
+
+static llvm::Function *lGetISPCIntrinsicsFuncDecl(llvm::Module *M, std::string origName, ISPCIntrinsics ID,
+                                                  std::vector<const Type *> &argTypes) {
+    std::string name = {};
+    llvm::Type *retType = nullptr;
+    std::vector<llvm::Type *> TYs;
+    for (const Type *type : argTypes) {
+        TYs.push_back(type->LLVMType(g->ctx));
+    }
+    bool hasUnnamedType = false;
+    switch (ID) {
+    case ISPCIntrinsics::atomicrmw: {
+        retType = TYs[1];
```

Comment:
Here and in similar cases below, please add validation for `TYs` size to avoid crashes

Reply:
Done